File: Stacks/Stacks.cpp
#include <iostream>
#include <stack>
#include <chrono>
#include <iomanip>

/**
 * Demonstrates stack operations with LIFO (Last In First Out) behavior.
 * Uses std::stack container adapter.
 * Pushes 5 integers and pops them to show reverse ordering.
 */
int main() {
    auto startTime = std::chrono::high_resolution_clock::now();
    
    // Create stack instance
    std::stack<int> myStack;

    // Push elements
    myStack.push(10);
    myStack.push(20);
    myStack.push(30);
    myStack.push(40);
    myStack.push(50);
    
    // Calculate memory (approximate for 5 integers)
    size_t memoryUsed = sizeof(int) * 5 + sizeof(std::stack<int>);

    // Pop and print elements
    std::cout << "Elements in the Stack:" << std::endl;
    while (!myStack.empty()) {
        std::cout << "Popped Element: " << myStack.top() << std::endl;
        myStack.pop();
    }
    
    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
    
    // Performance statistics
    std::cout << "\n--- Performance Statistics ---" << std::endl;
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "Execution time: " << duration.count() / 1000.0 << " ms" << std::endl;
    std::cout << "Memory usage: " << memoryUsed << " bytes" << std::endl;
    std::cout << "\nTime Complexity:" << std::endl;
    std::cout << "  - Push: O(1)" << std::endl;
    std::cout << "  - Pop: O(1)" << std::endl;
    std::cout << "  - Peek: O(1)" << std::endl;
    std::cout << "  - IsEmpty: O(1)" << std::endl;

    return 0;
}

File: Stacks/Stacks.java
package Stacks;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Demonstrates stack operations with LIFO (Last In First Out) behavior.
 * Uses Deque interface with ArrayDeque implementation.
 * Pushes 5 integers and pops them to show reverse ordering.
 */
public class Stacks {

    public static void main(String[] args) {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        long memoryBefore = runtime.totalMemory() - runtime.freeMemory();
        long startTime = System.nanoTime();
        
        // Create stack instance using ArrayDeque
        Deque<Integer> myStack = new ArrayDeque<>();

        // Push elements
        myStack.push(10);
        myStack.push(20);
        myStack.push(30);
        myStack.push(40);
        myStack.push(50);

        // Pop and print elements
        System.out.println("Elements in the Stack:");
        while (!myStack.isEmpty()) {
            System.out.println("Popped Element: " + myStack.pop());
        }
        
        long endTime = System.nanoTime();
        long memoryAfter = runtime.totalMemory() - runtime.freeMemory();
        long memoryUsed = memoryAfter - memoryBefore;
        
        // Performance statistics
        System.out.println("\n--- Performance Statistics ---");
        System.out.printf("Execution time: %.4f ms%n", (endTime - startTime) / 1_000_000.0);
        System.out.println("Memory usage: " + (memoryUsed > 0 ? memoryUsed : "negligible") + " bytes");
        System.out.println("\nTime Complexity:");
        System.out.println("  - Push: O(1)");
        System.out.println("  - Pop: O(1)");
        System.out.println("  - Peek: O(1)");
        System.out.println("  - IsEmpty: O(1)");
    }
}
